src/intrinsics/ecma262/ObjectPrototype.js

```diff
@@ -10,24 +10,45 @@
 /* @flow */
 
 import type { Realm } from "../../realm.js";
-import { NativeFunctionValue, ObjectValue, BooleanValue, NullValue } from "../../values/index.js";
+import { AbstractValue, NativeFunctionValue, ObjectValue, BooleanValue, NullValue } from "../../values/index.js";
 import { SameValuePartial, RequireObjectCoercible } from "../../methods/abstract.js";
 import { HasOwnProperty, HasSomeCompatibleType } from "../../methods/has.js";
 import { Invoke } from "../../methods/call.js";
 import { Properties, To } from "../../singletons.js";
+import { FatalError } from "../../errors.js";
+import type { BabelNodeExpression } from "babel-types";
+import * as t from "babel-types";
 import invariant from "../../invariant.js";
 
 export default function(realm: Realm, obj: ObjectValue): void {
   // ECMA262 19.1.3.2
-  obj.defineNativeMethod("hasOwnProperty", 1, (context, [V]) => {
-    // 1. Let P be ? ToPropertyKey(V).
-    let P = To.ToPropertyKey(realm, V.throwIfNotConcrete());
-
-    // 2. Let O be ? ToObject(this value).
-    let O = To.ToObjectPartial(realm, context);
-
-    // 3. Return ? HasOwnProperty(O, P).
-    return new BooleanValue(realm, HasOwnProperty(realm, O, P));
+  const ObjectPrototypeHasOwnPrototype = obj.defineNativeMethod("hasOwnProperty", 1, (context, [V]) => {
+    try {
+      // 1. Let P be ? ToPropertyKey(V).
+      let P = To.ToPropertyKey(realm, V.throwIfNotConcrete());
+
+      // 2. Let O be ? ToObject(this value).
+      let O = To.ToObjectPartial(realm, context);
+
+      // 3. Return ? HasOwnProperty(O, P).
+      return new BooleanValue(realm, HasOwnProperty(realm, O, P));
+    } catch (x) {
+      if (realm.isInPureScope() && x instanceof FatalError) {
```

Comment:
In effect, this is saying that realm.isInPureScope() implies that To.PropertyKey and To.ObjectPartial will not have observable side effects. This probably bears saying over and over again, lest people pick up on the try-catch pattern without realizing that there a huge and dangerous assumption needed before anything can be done in here.

Reply:
Ok. I'll add a comment.

Reply:
Actually, I'll hoist them out because I don't want to assume ToPropertyKey is pure since it's not.